pkg/sql/schemachanger/scbuild/internal/scbuildstmt/common_relation.go

```diff
@@ -17,70 +17,432 @@ import (
+	if dir == scpb.Target_ADD {
+		b.EnqueueAdd(elem)
+	} else {
+		b.EnqueueDropIfNotExists(elem)
 	}
 }
 
-// removeColumnTypeBackRefs removes type back references for a given table
-// column from default expressions and comptued expressions.
-func removeColumnTypeBackRefs(b BuildCtx, table catalog.TableDescriptor, id descpb.ColumnID) {
-	visitor := &tree.TypeCollectorVisitor{
-		OIDs: make(map[oid.Oid]struct{}),
+// columnDescToElement converts an individual column descriptor
+// into the equivalent column element.
+func columnDescToElement(
+	table catalog.TableDescriptor,
+	column descpb.ColumnDescriptor,
+	familyName *string,
+	familyID *descpb.FamilyID,
+) *scpb.Column {
+	nilToEmptyString := func(val *string) string {
+		if val == nil {
+			return ""
+		}
+		return *val
 	}
-	visitorDeleted := &tree.TypeCollectorVisitor{
-		OIDs: make(map[oid.Oid]struct{}),
+	// Find the column family entries, when they
+	// aren't populated
+	if familyID == nil &&
+		!column.Virtual {
+		columnFamilies := table.GetFamilies()
+		familyIdx := -1
+		for idx, family := range columnFamilies {
+			for _, col := range family.ColumnIDs {
+				if col == column.ID {
+					familyIdx = idx
+					break
+				}
+			}
+			if familyIdx != -1 {
+				break
+			}
+		}
+		familyID = &columnFamilies[familyIdx].ID
+		familyName = &columnFamilies[familyIdx].Name
+	} else if column.Virtual {
+		localID := descpb.FamilyID(0)
+		localName := ""
+		familyID = &localID
+		familyName = &localName
+	}
+
+	return &scpb.Column{
+		TableID:                           table.GetID(),
+		ColumnID:                          column.ID,
+		Type:                              column.Type,
+		FamilyName:                        *familyName,
+		FamilyID:                          *familyID,
+		Nullable:                          column.Nullable,
+		DefaultExpr:                       nilToEmptyString(column.DefaultExpr),
+		OnUpdateExpr:                      nilToEmptyString(column.OnUpdateExpr),
+		Hidden:                            column.Hidden,
+		Inaccessible:                      column.Inaccessible,
+		GeneratedAsIdentitySequenceOption: nilToEmptyString(column.GeneratedAsIdentitySequenceOption),
+		GeneratedAsIdentityType:           column.GeneratedAsIdentityType,
+		UsesSequenceIds:                   column.UsesSequenceIds,
+		ComputerExpr:                      nilToEmptyString(column.ComputeExpr),
+		PgAttributeNum:                    column.GetPGAttributeNum(),
+		SystemColumnKind:                  column.SystemColumnKind,
+		Virtual:                           column.Virtual,
 	}
-	// TODO(fqazi): Deal with the case where a column is added
-	// in the current statement.
+}
+
+type exprType int
 
+const (
+	exprTypeDefault  exprType = 0
+	exprTypeOnUpdate exprType = 1
+	exprTypeComputed exprType = 2
+)
+
+// decomposeExprToTypeRef converts and inserts type references from
+// an expression.
+func decomposeExprToElements(
```

Comment:
there's a few patterns in these decompositions I'm not sure about.

How much of the BuildCtx do we need? My sense is for most of it, not much. Fundamentally, I'm interested in asking whether we can define the decomposition without the tight coupling to the builder. I think it'll be valuable to perform full decompositions independently. I know it's a lot of packages but, what do you think about pulling the decomposition out into a package like `scelem`?

Second, the `addIfDuplicateDoesNotExistForDir` pattern smells to me.

Reply:
It'd be fun also to be able to test the decomposition stuff by starting with descriptors, decomposing them, and then running planning and execution to rebuild them and making sure that that still works.

Reply:
I'd be inclined to leave this code as it is and not worry about it, I'll be making a lot of changes to it soon anyway when working on the builder's view of the catalog state in terms of elements instead of descriptors.